core/templates/dev/head/services/UserService.js

```diff
@@ -19,16 +19,33 @@
 oppia.factory('UserService', [
   '$http', '$q', 'UrlInterpolationService', 'DEFAULT_PROFILE_IMAGE_PATH',
   function($http, $q, UrlInterpolationService, DEFAULT_PROFILE_IMAGE_PATH) {
-    var _isLoggedIn = GLOBALS.userIsLoggedIn;
     var PREFERENCES_DATA_URL = '/preferenceshandler/data';
 
+    var userInfo;
+
+    var getUserInfoAsync = function() {
+      if (GLOBALS.userIsLoggedIn) {
+        if (userInfo) {
+          return $q.resolve(userInfo);
+        }
+        return $http.get(
+          '/userinfohandler'
+        ).then(function(response) {
+          userInfo = response.data;
```

Comment:
Could we make a UserInfo domain object instead (with getters)? Better to have something conceptually-clear here, and also prevents deep binding to arbitrary dicts that get passed around the application.